### PR TITLE
#588: Make the constructors of Href compliant with ConstructorOnlyInitializesOrCallOtherConstructors

### PR DESCRIPTION
**Pull request by essobedo**
Fix for https://github.com/yegor256/takes/issues/588

The goal of this PR is to make the class Href compliant with the rule ConstructorOnlyInitializesOrCallOtherConstructors by delegating the construction of the map of query arguments and the clean up of the URI to a private static methods.

### CODE REVIEW

**Comment by davvd** [comment]
@essobedo Thanks, someone will review your pull request soon

**Comment by davvd** [comment]
@dmzaytsev it's yours, please review

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by essobedo** [comment]
@yegor256 please confirm

**Comment by essobedo** [comment]
@dmzaytsev PR updated to fix the conflicts, please review

File: src/main/java/org/takes/misc/Href.java [hunk 1]
```diff
@@ -321,4 +292,53 @@ private static URI createURI(final String txt) {
...
+     * Convert the provided query into a Map.
+     * @param query The query to parse.
+     * @return A map containing all the query arguments and their values.
+     */
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    private static SortedMap<String, List<String>> asMap(final String query) {
+        final SortedMap<String, List<String>> params =
+            new TreeMap<String, List<String>>();
+        if (query != null) {
+            final String[] pairs = query.split("&");
```
Comment (dmzaytsev) [inline src/main/java/org/takes/misc/Href.java]: @essobedo once used variable can be inline

**Comment by dmzaytsev** [comment]
@essobedo please see my comment above

File: src/main/java/org/takes/misc/Href.java (hunk 1 above)
Comment (essobedo) [inline src/main/java/org/takes/misc/Href.java]: @dmzaytsev I'm not sure I understand what you mean, could you please reformulate?

Comment (dmzaytsev) [inline src/main/java/org/takes/misc/Href.java]: @essobedo I mean
the variable `pairs` used only once
so this code
```
final String[] pairs = query.split("&");
    for (final String pair : pairs) {
```
can be rewritten as
```
    for (final String pair : query.split("&")) {
```

also please check out  this article 
http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#one-time-variables

**Comment by essobedo** [comment]
@dmzaytsev That's very clear now, please have a look to the new PR

**Comment by dmzaytsev** [comment]
@essobedo thank you

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by essobedo** [comment]
@yegor256 This PR is waiting for a confirmation for more than a week now, so could you please do something about it? confirm it or reject it but please do something about it.

**Comment by yegor256** [comment]
@rultor merge

**Comment by essobedo** [comment]
@yegor256 Thank you very much

**Comment by davvd** [comment]
@ypshenychka please, check this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by ypshenychka** [comment]
@dmzaytsev 
According to our [QA Rules](http://at.teamed.io/qa.html):
>The code reviewer found at least three problems in the code.

Only one issue was found during code review. 
Please confirm that you'll try to find at least three problems while future reviews.

**Comment by dmzaytsev** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@dmzaytsev Thanks.

**Comment by ypshenychka** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@ypshenychka thank you, let's try to make it "good" next time

**Comment by davvd** [comment]
@dmzaytsev paied 10 mins to @ypshenychka for QA review (payment ID is `79187912`). paid, thanks, added **22 mins** to your account, payment `79187922`, 382 hours and 14 mins was spent. extra minutes for review comments (c=7). +22 added to your rating, current score is: [+3714](http://www.netbout.com/b/35166?open=rating)

**Pull request merged**